### PR TITLE
Allow automatic discovery of OpenAPI spec files

### PR DESCRIPTION
**Pull request by sideeffffect**

### CODE REVIEW

File: src/sbt-test/sbt-guardrail/scala-client-codegen-app/build.sbt [hunk 1]
```diff
@@ -6,15 +6,20 @@ scalaVersion := "2.12.16"
...
+    ScalaServer(openApiFile.file, pkg = openApiFile.pkg + ".server", imports = List("_root_.support.PositiveLong"))
+  )
+}
+
+Test / guardrailTasks := (Test / guardrailDiscoveredOpenApiFiles).value.flatMap { openApiFile =>
+  List(
+    ScalaClient(openApiFile.file, pkg = openApiFile.pkg + ".client", imports = List("_root_.support.PositiveLong", "_root_.support.tests.PositiveLongInstances._")),
+    ScalaServer(openApiFile.file, pkg = openApiFile.pkg + ".server", imports = List("_root_.support.PositiveLong", "_root_.support.tests.PositiveLongInstances._"))
+  )
+}
```
Comment (blast-hardcheese) [inline src/sbt-test/sbt-guardrail/scala-client-codegen-app/build.sbt:21]: perhaps
```suggestion
Test / guardrailTasks += ScalaClient.defaults(imports = List(
    "_root_.support.PositiveLong",
    "_root_.support.tests.PositiveLongInstances._"
  ))
Test / guardrailTasks ++= (Test / guardrailDiscoveredOpenApiFiles).value.flatMap { openApiFile =>
  List(
    ScalaClient(openApiFile.file, pkg = openApiFile.pkg + ".client"),
    ScalaServer(openApiFile.file, pkg = openApiFile.pkg + ".server")
  )
}
```

is a useful example for those attempting to reduce duplication

**Review by blast-hardcheese — APPROVED** [review]
this is a really neat idea.

Initially I rejected the possibility of automatically generating clients and servers from files discovered in a directory because I thought more people would need to supply bespoke arguments for each generated client or server, but I like how this still gives control in a generic way.

Do you think it's going to be necessary to filter out files that have an explicit configuration stanza for them, for files that require custom arguments or other workarounds?

File: modules/core/src/main/scala/GuardrailHelpers.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      case l => l.dropRight(1)
+    }
+  }.mkString(".")
+
+  private def recursiveListFiles(f: File): List[File] = {
+    val these = Option(f.listFiles).toList.flatten
+    these.filter(f => f.isFile) ++ these.filter(_.isDirectory).flatMap(recursiveListFiles)
+  }
+
+  def isOpenApiSpec(file: File): Boolean = {
```
[the remark this answers is not in the document]
Comment (sideeffffect) [reply modules/core/src/main/scala/GuardrailHelpers.scala:21]: Yes, exactly. Not all YAML/JSON files under `src/main/openapi` are guaranteed to be full OpenAPI specs.

**Comment by sideeffffect** [comment]
> Do you think it's going to be necessary to filter out files that have an explicit configuration stanza for them, for files that require custom arguments or other workarounds?

I think the scheme is very flexible. You get a DiscoveredFile and are supposed to return a List of GuardRailTasks. That means you can inspect the file, return an empty list to filter it or anything in between.

**Comment by sideeffffect** [comment]
ping :smile_cat:

**Comment by blast-hardcheese** [comment]
@sideeffffect Hey, I was trying to figure out what didn't work about the `default` stuff. I use that functionality all the time with the CLI, I was really surprised to see it not work here, but that shouldn't block your PR. Sorry about that

**Comment by sideeffffect** [comment]
:heart:

**Pull request merged**